OSDOCS#9020: Add OLM v0 deprecations schema (#70898)
openshift/openshift-docs: modules/olm-deprecations-schema.adoc

// Module included in the following assemblies:
//
// * operators/understanding/olm-packaging-format.adoc

[id="olm-deprecations-schema_{context}"]
= olm.deprecations schema

The optional `olm.deprecations` schema defines deprecation information for packages, bundles, and channels in a catalog. Operator authors can use this schema to provide relevant messages about their Operators, such as support status and recommended upgrade paths, to users running those Operators from a catalog.

> An `olm.deprecations` schema entry can contain at least one of the following `reference` types, which indicates the deprecation scope. After the Operator is installed, any deprecation messages can be viewed as status conditions on the related `Subscription` object.

Reviewer: "deprecation messages" 
Do we need to be more explicit that it's `entries[].message` displayed as an opaque blob in status conditions for the subscription?

> An `olm.deprecations` schema entry can contain at least one of the following `reference` types, which indicates the deprecation scope. After the Operator is installed, any deprecation messages can be viewed as status conditions on the related `Subscription` object.

Reviewer: ```suggestion
An `olm.deprecations` schema entry can contain one or more of the following `reference` types, which indicates the deprecation scope. After the Operator is installed, any deprecation messages can be viewed as status conditions on the related `Subscription` object.
```

> An `olm.deprecations` schema entry can contain at least one of the following `reference` types, which indicates the deprecation scope. After the Operator is installed, any deprecation messages can be viewed as status conditions on the related `Subscription` object.

Author: ```suggestion
An `olm.deprecations` schema entry contains at least one of the following `reference` types, which indicates the deprecation scope. After the Operator is installed, any deprecation messages can be viewed as status conditions on the related `Subscription` object.
```
Think I'll go with this

.Deprecation `reference` types
[cols="1,2,1",options="header"]
|===
|Type |Scope |Status condition

|`olm.package`
|Represents the entire package
|`PackageDeprecated`

|`olm.channel`
|Represents one channel
|`ChannelDeprecated`

|`olm.bundle`
|Represents one bundle version
|`BundleDeprecated`

|===

Each `reference` type has their own requirements, as detailed in the following example.

[WARNING]
====

> If an `olm.deprecations` schema is determined invalid by the `opm validate` command, the entire catalog is invalid.

Reviewer: I'd have to look for the citation, but in general we should be saying something like "if any part of FBC is invalid, then all the FBC is considered invalid" and the deprecations stuff is just add'l FBC schema.

Reviewer: We dropped this whole warning in the upstream, as immaterial since FBC is either valid/invalid with no middle-ground.

====

.Example `olm.deprecations` schema with each `reference` type
[%collapsible]
====
[source,yaml]
----
schema: olm.deprecations <1>
package: my-operator <2>
entries:
  - reference:
      schema: olm.package <3>
    message: | <4>
    The 'my-operator' package is end of life. Please use the
    'my-operator-new' package for support.
  - reference:
      schema: olm.channel
      name: alpha <5>
    message: |
    The 'alpha' channel is no longer supported. Please switch to the
    'stable' channel.
  - reference:
      schema: olm.bundle
      name: my-operator.v1.68.0 <6>
    message: |
    my-operator.v1.68.0 is deprecated. Uninstall my-operator.v1.68.0 and
    install my-operator.v1.72.0 for support.
----
<1> There can be only one `olm.deprecations` schema per package.

> <2> The `package` name must be present in the bundle and unique across the catalog. There must not be an associated `name` field.

Reviewer: <1> and <2> are really the same thing.  The requirement is that each deprecation schema `must` have a package reference (<2>) and that must package reference be unique in the catalog (i.e. must be comprehensive w.r.t. all deprecation activities relevant to that package).  

For e.g. the following is valid:
```yaml
schema: olm.deprecations
package: foo
entries:
- reference
    schema: olm.channel
    name: alpha
  message: hey, bad channel, do <beta>
---
schema: olm.deprecations
package: bar
entries:
- reference: 
    schema: olm.bundle
    name: my-operator.v1.68.0
  message: dead version!  go back!
```
because there are two `olm.deprecations` in the catalog, but they relate to discrete packages.

<3> The `olm.package` schema must not include a `name` field, because it is determined by the `package` field already defined earlier in the schema.
<4> All `message` fields, for any `reference` type, must be a non-zero length and represented as an opaque text blob.
<5> The `name` field for the `olm.channel` schema is required.
<6> The `name` field for the `olm.bundle` schema is required.
====

[NOTE]
====

> The `olm.deprecations` schema does not consider overlapping deprecation, for example package versus channel versus bundle.

Reviewer: This is more a function of the underlying definition of FBC.  
Since FBC is represented by a directory hierarchy, so long as the schema is present in the aggregate and valid, then it's fine.  It is up to the author if they combine it with other schema or remains disjoint/discrete.

Reviewer: When I wrote this, I was going to ask for it to be struck from the upstream docs.  
We ended up keeping
```text
The deprecation feature does not consider overlapping deprecation (package vs channel vs bundle).
```
which (some flavor of) would be fine here.

====

> Operator authors can save their `olm.deprecations` schema entries as a `deprecations.yaml` file in the same directory as the package's `index.yaml` file:

Reviewer 2: Suggestion: You can remove "their".
